src/NuGet.Clients/NuGet.CommandLine/MsBuildToolset.cs

```diff
@@ -0,0 +1,173 @@
+                        _parsedToolsVersion = new Version(0, 0);
+                    }
+                }
+
+                return _parsedToolsVersion;
+            }
+        }
+
+        public string Version { get; private set; }
+
+        public string Path { get; private set; }
+
+        public int CompareTo(object obj)
+        {
+            if (Object.ReferenceEquals(obj, null))
+            {
+                return 1;
+            }
+
+            var rhs = obj as MsBuildToolset;
+            if (obj == null)
+            {
+                throw new ArgumentException("Comparison object not of correct type");
+            }
+
+            // Compare versions
+            var comparison = this.ParsedVersion.Major.CompareTo(rhs.ParsedVersion.Major);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Minor.CompareTo(rhs.ParsedVersion.Minor);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Build.CompareTo(rhs.ParsedVersion.Build);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            comparison = this.ParsedVersion.Revision.CompareTo(rhs.ParsedVersion.Revision);
+            if (comparison != 0)
+            {
+                return comparison;
+            }
+
+            // Versions equal; compare by install date/time
+            return this._installDate > rhs._installDate ? 1 : -1;
+        }
+
+        private static DateTime ConvertFILETIMEToDateTime(FILETIME time)
+        {
+            long highBits = time.dwHighDateTime;
+            highBits = highBits << 32;
+            return DateTime.FromFileTimeUtc(highBits | (long)(uint)time.dwLowDateTime);
+        }
+
+        public static string GetMsBuildDirFromVsDir(string vsDir)
+        {
+            if (string.IsNullOrEmpty(vsDir))
+            {
+                return null;
+            }
+
+            string msBuildRoot = System.IO.Path.Combine(vsDir, "MSBuild");
+            if (!Directory.Exists(msBuildRoot))
+            {
+                return null;
+            }
+
+            // Enumerate all versions of MSBuild present, take the highest
+            string msBuildDirectory = string.Empty;
+            var highestVersionRoot = Directory.EnumerateDirectories(msBuildRoot).OrderByDescending(dir =>
+            {
+                var dirName = new DirectoryInfo(dir).Name;
+                float dirValue;
+                if (float.TryParse(dirName, out dirValue))
+                {
+                    return dirValue;
+                }
+
+                return 0F;
+            })
+            .FirstOrDefault(dir =>
+            {
+                msBuildDirectory = System.IO.Path.Combine(dir, "bin");
+                return File.Exists(System.IO.Path.Combine(msBuildDirectory, "msbuild.exe"));
```

Comment:
We need to keep them separate as the first combine's result is the returned value. The second combine is for testing that a file exists in the returned path, which is the test for correctness.